channels/rdpecam/client/camera_device_main.c

```diff
@@ -0,0 +1,821 @@
+
+	return ecam_channel_write(dev->ecam, hchannel, msg, s, TRUE);
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT ecam_dev_process_sample_request(CameraDevice* dev, GENERIC_CHANNEL_CALLBACK* hchannel,
+                                            wStream* s)
+{
+	BYTE streamIndex = 0;
+
+	WINPR_ASSERT(dev);
+
+	if (!Stream_CheckAndLogRequiredLength(TAG, s, 1))
+		return ERROR_INVALID_DATA;
+
+	Stream_Read_UINT8(s, streamIndex);
+
+	if (streamIndex >= ECAM_DEVICE_MAX_STREAMS)
+	{
+		WLog_ERR(TAG, "Incorrect streamIndex %d", streamIndex);
+		ecam_channel_send_error_response(dev->ecam, hchannel, CAM_ERROR_CODE_InvalidStreamNumber);
+		return ERROR_INVALID_INDEX;
+	}
+
+	CameraDeviceStream* stream = &dev->streams[streamIndex];
+
+	/* need to save channel because responses are asynchronous and coming from capture thread */
+	if (stream->hSampleReqChannel != hchannel)
+		stream->hSampleReqChannel = hchannel;
+
+	/* allow to send that many unsolicited samples */
+	stream->nSampleCredits = ECAM_MAX_SAMPLE_CREDITS;
+
+	return CHANNEL_RC_OK;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT ecam_dev_process_current_media_type_request(CameraDevice* dev,
+                                                        GENERIC_CHANNEL_CALLBACK* hchannel,
+                                                        wStream* s)
+{
+	BYTE streamIndex = 0;
+
+	WINPR_ASSERT(dev);
+
+	if (!Stream_CheckAndLogRequiredLength(TAG, s, 1))
+		return ERROR_INVALID_DATA;
+
+	Stream_Read_UINT8(s, streamIndex);
+
+	if (streamIndex >= ECAM_DEVICE_MAX_STREAMS)
+	{
+		WLog_ERR(TAG, "Incorrect streamIndex %d", streamIndex);
+		ecam_channel_send_error_response(dev->ecam, hchannel, CAM_ERROR_CODE_InvalidStreamNumber);
+		return ERROR_INVALID_INDEX;
+	}
+
+	CameraDeviceStream* stream = &dev->streams[streamIndex];
+
+	if (stream->currMediaType.Format == 0)
+	{
+		WLog_ERR(TAG, "Current media type unknown for streamIndex %d", streamIndex);
+		ecam_channel_send_error_response(dev->ecam, hchannel, CAM_ERROR_CODE_NotInitialized);
+		return ERROR_DEVICE_REINITIALIZATION_NEEDED;
+	}
+
+	return ecam_dev_send_current_media_type_response(dev, hchannel, &stream->currMediaType);
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT ecam_dev_send_media_type_list_response(CameraDevice* dev,
+                                                   GENERIC_CHANNEL_CALLBACK* hchannel,
+                                                   CAM_MEDIA_TYPE_DESCRIPTION* mediaTypes,
+                                                   size_t nMediaTypes)
+{
+	CAM_MSG_ID msg = CAM_MSG_ID_MediaTypeListResponse;
+
+	WINPR_ASSERT(dev);
+
+	wStream* s = Stream_New(NULL, CAM_HEADER_SIZE + ECAM_MAX_MEDIA_TYPE_DESCRIPTORS *
+	                                                    sizeof(CAM_MEDIA_TYPE_DESCRIPTION));
```

Comment:
The use of sizeof here is a bad idea, if there's some alignment `sizeof(CAM_MEDIA_TYPE_DESCRIPTION)` will probably not return what we expect.

Reply:
But we don't care in that case if the allocation is a little too big